### PR TITLE
feat: added API url configuration for Ollama and Hugging Face models

### PR DESCRIPTION
**Pull request by utkuaydn**
### **User description**
I hope the title is self-describing. It would be great if this feature is available since LiteLLM allows the usage of custom API URLs. Any comments or reviews are welcome.

___

### **PR Type**
Enhancement, Documentation

___

### **Description**
- Added `api_base` parameter to `AICaller` class to support custom API URLs for Ollama and Hugging Face models.
- Updated `generate_tests` method in `UnitTestGenerator` to accept `API_base` parameter.
- Added `--api-base` argument to the command-line interface for specifying custom API URLs.
- Updated tests for `AICaller` to include `api_base` parameter.
- Documented the usage of custom API URLs for Ollama and Hugging Face models in the README.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>AICaller.py</strong><dd><code>Add API base URL configuration to AICaller class</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/AICaller.py

<li>Added <code>api_base</code> parameter to <code>AICaller</code> class.<br> <li> Modified <code>call_model</code> method to use <code>api_base</code> for Ollama and Hugging Face <br>models.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/27/files#diff-d94557e610fe5a4d747c06e8ac8d6a0e243e5c50b4df8f624fa1acc7d51f57b0">+4/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>UnitTestGenerator.py</strong><dd><code>Include API base URL in UnitTestGenerator</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/UnitTestGenerator.py

<li>Added <code>API_base</code> parameter to <code>generate_tests</code> method.<br> <li> Updated instantiation of <code>AICaller</code> with <code>API_base</code>.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/27/files#diff-19760582d9ede3a799fdbb541ad357b4822682e837bca8365196fba50daf57e3">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>main.py</strong><dd><code>Add API base URL argument to CLI</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
cover_agent/main.py

- Added `--api-base` argument to command-line interface.

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/27/files#diff-610f77abc2025129e317ec1f98b3bb375b790b2197ed3f635331cca0a06fd999">+5/-0</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Tests
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>test_AICaller.py</strong><dd><code>Update AICaller tests to include API base URL</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
tests/test_AICaller.py

<li>Updated tests to include <code>api_base</code> parameter in <code>AICaller</code> instantiation.<br> <br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/27/files#diff-e84aeed71db4a60462435bc58814bc055ba7779343c529c6531763006007a961">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></td></tr><tr><td><strong>Documentation
</strong></td><td><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Document custom API URL usage for specific models</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>
      
README.md

<li>Added documentation for using custom API URLs with Ollama and Hugging <br>Face models.<br>

</details>
    

  </td>
  <td><a href="https://github.com/Codium-ai/cover-agent/pull/27/files#diff-b335630551682c19a781afebcf4d07bf978fb1f8ac04c6bf87428ed5106870f5">+13/-0</a>&nbsp; &nbsp; </td>

</tr>                    
</table></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: cover_agent/AICaller.py [hunk 1]
```diff
@@ -23,7 +25,7 @@ def call_model(self, prompt, max_tokens=4096):
             tuple: A tuple containing the response generated by the language model, the number of tokens used from the prompt, and the total number of tokens in the response.
         """
         messages = [{"role": "user", "content": prompt}]
-        response = litellm.completion(model=self.model, messages=messages, max_tokens=max_tokens, stream=True)
+        response = litellm.completion(model=self.model, api_base=self.api_base, messages=messages, max_tokens=max_tokens, stream=True) if "ollama" in self.model or "huggingface" in self.model else litellm.completion(model=self.model, messages=messages, max_tokens=max_tokens, stream=True)
```
Comment (EmbeddedDevops1) [inline cover_agent/AICaller.py:28]: While this is definitely _Pythonic_ I feel like this calls runs a bit too long. Any chance we can break this apart into a multi-line `if` statement for readability purposes?

File: cover_agent/AICaller.py [hunk 2]
```diff
@@ -2,14 +2,16 @@
 import litellm
 
 class AICaller:
-    def __init__(self, model):
+    def __init__(self, model, api_base):
```
Comment (EmbeddedDevops1) [inline cover_agent/AICaller.py:5]:
```suggestion
    def __init__(self, model: str, api_base: str):
```

File: cover_agent/UnitTestGenerator.py [hunk 3]
```diff
@@ -160,9 +160,9 @@ def build_prompt(self):
 
         return prompt.build_prompt()
 
-    def generate_tests(self, LLM_model="gpt-4o", max_tokens=4096, dry_run=False):
+    def generate_tests(self, LLM_model="gpt-4o", API_base="http://localhost:11434", max_tokens=4096, dry_run=False):
```
Comment (EmbeddedDevops1) [inline cover_agent/UnitTestGenerator.py:163]: I don't believe the intent was to hardcode a base URL into the script. Correct me if I'm wrong.
```suggestion
    def generate_tests(self, LLM_model="gpt-4o", max_tokens=4096, dry_run=False):
```

Comment (EmbeddedDevops1) [reply cover_agent/UnitTestGenerator.py:163]: I see the problem here. I'm going to refactor on your branch to move the AICaller instantiation to happen at the init phase of this class. Glad you surfaced this issue.

**Comment by EmbeddedDevops1** [comment]
@utkuaydn Thanks again for putting this PR together. Since I needed to refactor I added the changes from your commit into the refactor (cited you in the commit log). Are you able to checkout my branch and test it with your local models?

PR Link: https://github.com/Codium-ai/cover-agent/pull/29

**Comment by utkuaydn** [comment]
Hi @EmbeddedDevops1. So, I've tried running
```
poetry run cover-agent \
  --source-file-path "cover_agent/main.py" \
  --test-file-path "tests/test_main.py" \
  --code-coverage-report-path "coverage.xml" \
  --test-command "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO" \
  --coverage-type "cobertura" \
  --desired-coverage 70 \
  --max-iterations 1 \
  --model "ollama/codeqwen:chat" \
  --api-base "127.0.0.1:11434"
```
however, I believe there are other problems now. This is the stack trace:
```
2024-05-23 09:54:04,282 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO"
2024-05-23 09:54:13,567 - cover_agent.main - INFO - Current Coverage: 61.4%
2024-05-23 09:54:13,567 - cover_agent.main - INFO - Desired Coverage: 70%

Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = OpenAI(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise OpenAIError(
openai.OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 1079, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 1052, in completion
    response = openai_chat_completions.completion(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 582, in completion
    raise OpenAIError(status_code=500, message=traceback.format_exc())
litellm.llms.openai.OpenAIError: Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = OpenAI(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise OpenAIError(
openai.OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable


During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/utku/cover-agent/cover_agent/main.py", line 150, in main
    generated_tests = test_gen.generate_tests(max_tokens=4096)
  File "/home/utku/cover-agent/cover_agent/UnitTestGenerator.py", line 180, in generate_tests
    response, prompt_token_count, response_token_count = self.ai_caller.call_model(prompt=self.prompt, max_tokens=max_tokens)
  File "/home/utku/cover-agent/cover_agent/AICaller.py", line 33, in call_model
    response = litellm.completion(model=self.model, messages=messages, max_tokens=max_tokens, stream=True)
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 3415, in wrapper
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 3308, in wrapper
    result = original_function(*args, **kwargs)
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 2370, in completion
    raise exception_type(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 9885, in exception_type
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 8551, in exception_type
    raise AuthenticationError(
litellm.exceptions.AuthenticationError: OpenAIException - Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = openai(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise openaiError(
openai.openaiError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable

2024-05-23 09:54:14,689 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO"
2024-05-23 09:54:23,675 - cover_agent.main - INFO - Current Coverage: 61.4%
2024-05-23 09:54:23,676 - cover_agent.main - INFO - Desired Coverage: 70%
Streaming results from LLM model...
Error during streaming: Request URL is missing an 'http://' or 'https://' protocol.


Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/utku/cover-agent/cover_agent/main.py", line 150, in main
    generated_tests = test_gen.generate_tests(max_tokens=4096)
  File "/home/utku/cover-agent/cover_agent/UnitTestGenerator.py", line 180, in generate_tests
    response, prompt_token_count, response_token_count = self.ai_caller.call_model(prompt=self.prompt, max_tokens=max_tokens)
  File "/home/utku/cover-agent/cover_agent/AICaller.py", line 46, in call_model
    model_response = litellm.stream_chunk_builder(chunks, messages=messages)
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 4273, in stream_chunk_builder
    if chunks[0]._hidden_params.get("created_at", None):
IndexError: list index out of range
```

**Comment by utkuaydn** [comment]
Hi @EmbeddedDevops1. So, I checked out to master and tried running
```
poetry run cover-agent \
  --source-file-path "cover_agent/main.py" \
  --test-file-path "tests/test_main.py" \
  --code-coverage-report-path "coverage.xml" \
  --test-command "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO" \
  --coverage-type "cobertura" \
  --desired-coverage 70 \
  --max-iterations 1 \
  --model "ollama/codeqwen:chat" \
  --api-base "http://127.0.0.1:11434"
```
Even though it seems to have worked, there are still exceptions being thrown. This is the stack trace:
```
2024-05-23 10:59:53,033 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO"
2024-05-23 11:00:02,267 - cover_agent.main - INFO - Current Coverage: 61.4%
2024-05-23 11:00:02,267 - cover_agent.main - INFO - Desired Coverage: 70%

Give Feedback / Get Help: https://github.com/BerriAI/litellm/issues/new
LiteLLM.Info: If you need to debug this error, use `litellm.set_verbose=True'.

Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = OpenAI(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise OpenAIError(
openai.OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 1079, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 1052, in completion
    response = openai_chat_completions.completion(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 582, in completion
    raise OpenAIError(status_code=500, message=traceback.format_exc())
litellm.llms.openai.OpenAIError: Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = OpenAI(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise OpenAIError(
openai.OpenAIError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "/home/utku/cover-agent/cover_agent/main.py", line 150, in main
    generated_tests = test_gen.generate_tests(max_tokens=4096)
  File "/home/utku/cover-agent/cover_agent/UnitTestGenerator.py", line 180, in generate_tests
    response, prompt_token_count, response_token_count = self.ai_caller.call_model(prompt=self.prompt, max_tokens=max_tokens)
  File "/home/utku/cover-agent/cover_agent/AICaller.py", line 33, in call_model
    response = litellm.completion(model=self.model, messages=messages, max_tokens=max_tokens, stream=True)
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 3415, in wrapper
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 3308, in wrapper
    result = original_function(*args, **kwargs)
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/main.py", line 2370, in completion
    raise exception_type(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 9885, in exception_type
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/utils.py", line 8551, in exception_type
    raise AuthenticationError(
litellm.exceptions.AuthenticationError: OpenAIException - Traceback (most recent call last):
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 574, in completion
    raise e
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 492, in completion
    return self.streaming(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/litellm/llms/openai.py", line 654, in streaming
    openai_client = openai(
  File "/home/utku/.cache/pypoetry/virtualenvs/cover-agent-_Cne6EjI-py3.10/lib/python3.10/site-packages/openai/_client.py", line 104, in __init__
    raise openaiError(
openai.openaiError: The api_key client option must be set either by passing api_key to the client or by setting the OPENAI_API_KEY environment variable

2024-05-23 11:00:03,392 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO"
2024-05-23 11:00:12,520 - cover_agent.main - INFO - Current Coverage: 61.4%
2024-05-23 11:00:12,521 - cover_agent.main - INFO - Desired Coverage: 70%
Streaming results from LLM model...
@patch("cover_agent.main.os.path.isfile")
@patch("cover_agent.main.UnitTestGenerator")
def test_main_coverage_report_not_found(mock_unit_cover_agent, mock_isfile):
    # Mocking argparse.Namespace object
    args = argparse.Namespace(
        source_file_path="test_source.py",
        test_file_path="test_file.py",
        code_coverage_report_path="coverage_report.xml",
        test_command="pytest",
        test_command_dir=os.getcwd(),
        included_files=None,
        coverage_type="cobertura",
        report_filepath="test_results.html",
        desired_coverage=90,
        max_iterations=10,
    )
    parse_args = lambda: args  # Mocking parse_args function
    mock_isfile.return_value = False  # Simulate coverage report not found

    with patch("cover_agent.main.parse_args", parse_args):
        with pytest.raises(FileNotFoundError) as exc_info:
            main()

    # Assert that FileNotFoundError was raised with the correct message
    assert (
        str(exc_info.value) == f"Coverage report not found at {args.code_coverage_report_path}"
    )

    # Assert that UnitTestGenerator was not called
    mock_unit_cover_agent.assert_not_called()

2024-05-23 11:00:22,272 - cover_agent.UnitTestGenerator - INFO - Total token used count for LLM model ollama/codeqwen:chat: 2928
2024-05-23 11:00:22,273 - cover_agent.UnitTestGenerator - INFO - Running test with the following command: "poetry run pytest --junitxml=testLog.xml --cov=templated_tests --cov=cover_agent --cov-report=xml --cov-report=term --log-cli-level=INFO"
2024-05-23 11:00:31,412 - cover_agent.UnitTestGenerator - ERROR - Test failed. Rolling back
2024-05-23 11:00:31,412 - cover_agent.main - INFO - Reached maximum iteration limit without achieving desired coverage.
```

**Comment by EmbeddedDevops1** [comment]
@utkuaydn Thanks so much for testing out my change. I was able to get it working with Ollama locally on my machine. I added [this section](https://github.com/Codium-ai/cover-agent?tab=readme-ov-file#using-other-llms) to the README that instructs you to add environment variables the same way LiteLLM works. For Ollama, however, you shouldn't need to export any environment variables so that's really weird.

## Configuration
Here's my setup:
* No environment variable set for OpenAI (i.e. `unset OPENAI_API_KEY`)
* Ollama installation environment: Latest version installed in Ubuntu 22.04 via WSL 2
* Graphics Card: NVIDIA GeForce GTX 1660

### Run command
```
$ poetry run cover-agent   --source-file-path "templated_tests/python_fastapi/app.py"   --test-file-path "templated_tests/python_fastapi/test_app.py"   --code-coverage-report-path "templated_tests/python_fastapi/coverage.xml"   --test-command "pytest --cov=. --cov-report=xml --cov-report=term"   --test-command-dir "templated_tests/python_fastapi"   --coverage-type "cobertura"   --desired-coverage 70   --max-iterations 10   --model "ollama/codellama:13b"
2024-05-23 10:48:06,978 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "pytest --cov=. --cov-report=xml --cov-report=term"
```

### Output

> 2024-05-23 10:48:06,978 - cover_agent.UnitTestGenerator - INFO - Running initial build/test command to generate coverage report: "pytest --cov=. --cov-report=xml --cov-report=term"
> 2024-05-23 10:48:07,811 - cover_agent.main - INFO - Current Coverage: 60.47%
> 2024-05-23 10:48:07,811 - cover_agent.main - INFO - Desired Coverage: 70%
> Streaming results from LLM model...
> 
> [PYTHON]
> ```
> def test_add_positive_numbers():
>     """
>     Test the add() function with positive numbers.
>     """
>     num1 = 5
>     num2 = 7
>     expected_result = 12
>     actual_result = add(num1, num2)
>     assert actual_result == expected_result
> ```
> 
> def test_add_negative_numbers():
>     """
>     Test the add() function with negative numbers.
>     """
>     num1 = -5
>     num2 = -7
>     expected_result = -12
>     actual_result = add(num1, num2)
>     assert actual_result == expected_result
> ```
> def test_add_positive_and_negative_numbers():
>     """
>     Test the add() function with a positive and negative number.
>     """
>     num1 = 5
>     num2 = -7
>     expected_result = -2
>     actual_result = add(num1, num2)
>     assert actual_result == expected_result
> ```

Assuming you're running in Ubuntu (or Debian?) and you've configured Ollama as a service you can tail the Ollama log and make sure you're hitting the `api/generate` endpoint from Cover Agent:
```
sudo journalctl -u ollama.service -f
```

Lastly, did you run just a straight Curl command against Ollama to validate all connections are good? For example,
```
curl http://127.0.0.1:11434/api/generate -d '{
  "model": "mistral:latest",
  "prompt": "Enter your text generation prompt here",
  "max_tokens": 100
}'
```

If so, have you validated that LiteLLM can talk to your Ollama server as well?

**Comment by EmbeddedDevops1** [comment]
I also just tested using a Docker container (`python:latest`) using `http://host.docker.internal:11434` as my `--api-base` and that worked as well.

**Comment by utkuaydn** [comment]
Thanks for the tips @EmbeddedDevops1. I can see in Ollama logs that the endpoints are for sure getting hits. 
```
[GIN] 2024/05/27 - 14:30:17 | 200 |   15.3675464s |       127.0.0.1 | POST     "/api/generate"
```
I tried running cover-agent with the flags you mentioned above but most of the tests that are generated fail with 
```
2024-05-27 14:42:44,409 - cover_agent.UnitTestGenerator - INFO - Skipping a generated test that failed
2024-05-27 14:42:44,409 - cover_agent.UnitTestGenerator - INFO - Running test with the following command: "pytest --cov=. --cov-report=xml --cov-report=term"
```

And, at the end of the execution, I get
```2024-05-27 14:42:45,907 - cover_agent.main - INFO - Reached maximum iteration limit without achieving desired coverage.```

Is this also the result you get?

**Comment by EmbeddedDevops1** [comment]
@utkuaydn Try with the latest. @mrT23 made some really good iterations on the prompting and parsing side of things. I'd also recommend starting with `ollama/mistral` first as a baseline.

It could be your LLM isn't following the exact ruleset when returning the code. If that's the case feel free to share feedback in the issue and make sure to paste the streamed tests that it generated (along with the LLM name).

Thanks again for your contribution. I'm going to close this PR for now but please feel free to open an issue or another PR as you see fitting.

**Pull request closed without merging**